ingest/src/main/java/com/microsoft/azure/kusto/ingest/ResourceAlgorithms.java

```diff
@@ -0,0 +1,136 @@
+package com.microsoft.azure.kusto.ingest;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.microsoft.azure.kusto.data.Utils;
+import com.microsoft.azure.kusto.data.instrumentation.FunctionOneException;
+import com.microsoft.azure.kusto.data.instrumentation.MonitoredActivity;
+import com.microsoft.azure.kusto.data.instrumentation.Tracer;
+import com.microsoft.azure.kusto.ingest.exceptions.IngestionClientException;
+import com.microsoft.azure.kusto.ingest.exceptions.IngestionServiceException;
+import com.microsoft.azure.kusto.ingest.resources.RankedStorageAccount;
+import com.microsoft.azure.kusto.ingest.resources.ResourceWithSas;
+import com.microsoft.azure.kusto.ingest.utils.SecurityUtils;
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.InputStream;
+import java.lang.invoke.MethodHandles;
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public class ResourceAlgorithms {
+    private static final int RETRY_COUNT = 3;
+    private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+    private ResourceAlgorithms() {
+    }
+
+    private static <TInner, TWrapper extends ResourceWithSas<TInner>, TOut> TOut resourceActionWithRetries(ResourceManager resourceManager,
+            List<TWrapper> resources, FunctionOneException<TOut, TWrapper, Exception> action, String actionName, Map<String, String> additionalAttributes)
+            throws IngestionClientException {
+
+        if (resources.isEmpty()) {
```

Comment:
Can you use Apache Commons' `ArrayUtils`?

Reply:
What's the advantage over using a native method?

Reply:
Includes `null` check (not sure if relevant here)

Reply:
Arrayutils doesn't work since it's a list, and we currently don't have the ListUtils dependency.